### PR TITLE
US1198692: Instrument-based repeated capabilities for attributes

### PR DESCRIPTION
**Pull request by luisgomes252**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/nimi-python/blob/master/CONTRIBUTING.md).
- [ ] ~~I've updated [CHANGELOG.md](https://github.com/ni/nimi-python/blob/master/CHANGELOG.md) if applicable.~~
- [x] I've added tests applicable for this pull request

### What does this Pull Request accomplish?

1. Add support for instrument-based repeated capabilities in attributes
2. Rename the instrument-based capability from 'devices' to 'instruments'
3. Update metadata properties to be more consistent across the internal code generation tools.

### What testing has been done?

Unit test for internal function, plus visual inspection of code-generated API elements

### CODE REVIEW

**Comment by sbethur** [comment]
FYI - @luisgomes252, There are merge conflicts.

**Comment by sbethur** [comment]
ok to test

**Comment by sbethur** [comment]
test this please

File: src/nidcpower/metadata/attributes.py
Comment (sbethur) [inline src/nidcpower/metadata/attributes.py:37]: This doesn't support `instruments` rep cap?

More broadly, are the supported rep cap changes already reviewed in internal metadata?

File: build/helper/documentation_helper.py
Comment (marcoskirsch) [inline build/helper/documentation_helper.py:875]: please add docstring

File: docs/nidcpower/rep_caps.rst
Comment (marcoskirsch) [inline docs/nidcpower/rep_caps.rst:45]: I think we need to revisit this:

1. '0-2' is not a goo example for `instruments` rep cap
2. `channel_enabled` makes no sense to be set on `instruments`
3. "passes a string of :python:`'0, 1, 2'` to the set attribute function" is irrelevant to Python programmers.

I do believe it should be done separately. Once this is merged please open an issue about this documentation problem.

File: src/nidcpower/metadata/attributes.py
Comment (marcoskirsch) [inline src/nidcpower/metadata/attributes.py:83]: Is the nimi-python metadata expansion defaulting lack of `supported_rep_caps` to mean `channels`?
I think that's a bad idea. We should make it so that the nimi-python metadata is explicit. 
Lack of `supported_rep_caps` should mean "None".

**Review by marcoskirsch — CHANGES_REQUESTED** [review]

File: build/helper/documentation_helper.py
Comment (sbethur) [inline build/helper/documentation_helper.py:126]: This key is used in [nidigital and nifgen](https://github.com/search?q=repeated_capability_type+repo%3Ani%2Fnimi-python+filename%3Aattributes.py&type=Code&ref=advsearch&l=&l=). I'm surprised changing this didn't update their generated code and docs.

File: build/helper/documentation_helper.py
Comment (sbethur) [inline build/helper/documentation_helper.py:197]: Can you please fix this existing mistake - the method docstring is not accurate.

File: src/nidcpower/metadata/attributes.py
Comment (sbethur) [reply src/nidcpower/metadata/attributes.py:83]: > Is the nimi-python metadata expansion defaulting lack of supported_rep_caps to mean channels?

I don't think this is true, but I'll let Luis confirm.

File: build/helper/documentation_helper.py
Comment (luisgomes252) [reply build/helper/documentation_helper.py:129]: Good idea. I've done that now.

File: build/helper/documentation_helper.py
Comment (luisgomes252) [reply build/helper/documentation_helper.py:877]: Yeah, you're right of course. I'll change it.

File: src/nidcpower/metadata/attributes.py
Comment (luisgomes252) [reply src/nidcpower/metadata/attributes.py:83]: Shreyas is correct. That's not true. The lack of `supported_rep_caps` means "None". That's why all those `'channel_based': False` attributes simply disappeared without having to be replaced by anything.

File: docs/nidcpower/rep_caps.rst
Comment (luisgomes252) [reply docs/nidcpower/rep_caps.rst:45]: #1602

File: build/helper/documentation_helper.py
Comment (luisgomes252) [reply build/helper/documentation_helper.py:126]: It did. I manually removed the generated files of the other drivers from the commit.

Even though I've already updated the metadata for the other drivers, I didn't include the output from the metadata exports of the other drivers (e.g. attributes.py) as inputs to the code generation in this repo, and so the code generated files would have been incorrect.

Comment (sbethur) [reply build/helper/documentation_helper.py:126]: Ok, Please have a work item in your backlog to update the metadata and generated files for those drivers.

**Comment by sbethur** [comment]
test this please

**Review by sbethur — CHANGES_REQUESTED** [review]
I don't have any further comments. Once you address all my comments, please re-request review.

**Comment by sbethur** [comment]
Fix for the Travis CI failure: https://github.com/ni/nimi-python/pull/1603

**Comment by sbethur** [comment]
test this please

File: generated/nidcpower/nidcpower/session.py
Comment (marcoskirsch) [inline generated/nidcpower/nidcpower/session.py:138]: This text is pretty old and not great. Here's a stab at improving it to better reflect the state of the art in our Python APIs and to be clearer to customers.

```
This property can be set/get to specific channels within you nidcpower.Session object. Use Python index notation on the repeated capabilities container `channels` to specify a subset.
Example: `my_session.channels[ ... ].active_advanced_sequence_step`,
To set/get on all, you can call the property directly on the nidcpower.Session:
`my_session.active_advanced_sequence_step`.
```

Comment (marcoskirsch) [reply generated/nidcpower/nidcpower/session.py:138]: #1604

**Review by marcoskirsch — CHANGES_REQUESTED** [review]

File: build/helper/documentation_helper.py
Comment (marcoskirsch) [reply build/helper/documentation_helper.py:126]: Does this mean that should someone clone the repo and run codegen would see generated files that are different from what was committed? That's not a state we want to leave the repo in. (unresolved until clarification)

File: generated/nidcpower/nidcpower/session.py
Comment (luisgomes252) [reply generated/nidcpower/nidcpower/session.py:138]: 1. @marcoskirsch, does the fact that you created an issue imply that you'd rather I not switch to using your text in this PR?
2. The current text is parameterized, so that the "...for channels" part, in the first line, can change with the particular capability type(s). Does your new text also allow for that? Specifically, if I were to replace "... to specific **channels**" with, say, "... to specific **channels, instruments and pins**", would that still be okay?
3. As for the example in the second line, I could replace it with the first item in the types list (allowing for cases where an attribute is valid for more than one type), since I can't use multiple types in the example. But I suppose I'd also have to pick an attribute that makes sense for that capability type. That's going to make it a bit harder to parameterize this text, as @sbethur [requested](https://github.com/ni/nimi-python/pull/1599#discussion_r630524419) earlier in this PR.

File: build/helper/documentation_helper.py
Comment (luisgomes252) [reply build/helper/documentation_helper.py:1317]: We were. I've added one now.

File: build/helper/documentation_helper.py
Comment (luisgomes252) [reply build/helper/documentation_helper.py:126]: Yes, @marcoskirsch, you are correct. Your scenario would happen as you described.

I left out the digital, dmm, fgen, modinst, scope, switch and tclk files from this PR because I assumed we wouldn't want to include diffs in this PR belonging to drivers, unrelated to the issue being addressed by this PR, and also because the internal nidigital build has been broken for a while (not sure it if still is) and so it wouldn't have been possible for me to use the metadata changes that I made for those drivers in the codegen build for this PR. Until that happens, I can't generate correct versions of at least the digital and scope files.

So now I see 3 possibilities. Let me know which one you prefer:

1. Block this PR until the nidigital build is fixed, and then include all the other updated driver files as part of this PR
2. Submit this PR as is, wait for the nidigital build to be fixed, and then immediately implement the user story that @sbethur ask that I create, which captured the task of updating those other files. This will result in the repo having latent, pending changes as you pointed out, at least for a couple of days, until the next PR goes through.
3. Include in this PR the version of the files that are currently being generated, which are incorrect for digital and scope. Similarly to item 2, this would be fixed in a few days. It's only the generated documentation that is incorrect, so maybe that's not so terrible.

I had implicitly chosen option 2, based on the fact that changes have gone into this repo in past that consisted of doing direct edits on the generated files, which also triggers the same scenario that you described. And in this case, we'd be in that state for a shorter amount of time. But if you guys prefer one of the other options, or know of a different option, let me know, and I'll be happy to comply.

Comment (marcoskirsch) [reply build/helper/documentation_helper.py:126]: > based on the fact that changes have gone into this repo in past that consisted of doing direct edits on the generated files

Really? That is pretty bad. Or are you talking about the metadata files generated internally at NI and uploaded to GitHub?

The problem with committing like this is that the files will show up as noise to every other developer.
@sbethur discussed and we think option 3 is best. Documentation will be temporarily wrong but is possibly the lesser of 3 evils.

File: generated/nidcpower/nidcpower/session.py
Comment (marcoskirsch) [reply generated/nidcpower/nidcpower/session.py:138]: 1. own PR, that's why I created the issue to track
2. yes
3. 👍🏻

File: build/helper/documentation_helper.py
Comment (luisgomes252) [reply build/helper/documentation_helper.py:126]: Sorry, yes, I was referring to editing the metadata files (functions.py, etc...), rather than changing their source hapigen files. So I guess cloning the repo and rebuilding wouldn't have resulted in changes. Sorry for conflating the two.

**Comment by sbethur** [comment]
test this please

File: docs/nitclk/class.rst
Comment (marcoskirsch) [inline docs/nitclk/class.rst:471]: There is no concept of Repeated Capabilities in `nitclk`.
Or `nimodinst` for that matter.
NI-DMM has the concept but it isn't applicable to the `nidmm` Python module.

I'm thinking we might be better off not even displaying this row at all for APIs without repeated capabilities.

I think Resettable, while a valid thing, should not be here either because the Python APIs have not exposed it (most driver runtimes don't implement it I believe)

#1606

**Review by marcoskirsch — APPROVED** [review]

**Review by sbethur — APPROVED** [review]

**Pull request merged**
